### ISSUE TITLE
Reason is lost when returning synth from pipe

### ISSUE THREAD

**Issue by gquintard** [varnishcache/varnish-cache#3329]
This one sorts of rings a bell, and it may be known behavior, but it definitely caught me off guard:

```
varnishtest "Lost reason from pipe to synth"

server s1 {} -start

varnish v1 -vcl+backend {
	sub vcl_recv {
		return (pipe);
	}
	sub vcl_pipe {
		return (synth(505, req.http.foo + "/bar"));
	}
} -start

client c1 {
	txreq -hdr "foo: foo"
	rxresp
	expect resp.reason == "foo/bar"
} -run
```

gives me:

```
**** v1   vsl|       1001 RespProtocol    c HTTP/1.1
**** v1   vsl|       1001 RespStatus      c 505
**** v1   vsl|       1001 RespReason      c 
```

**Comment by Dridi** [comment]
It only happens with complex expressions, probably because the result is allocated on the busy object's workspace while `vcl_synth` operates on the request workspace:

```vcl
varnishtest "(not) Lost reason from pipe to synth"

varnish v1 -vcl+backend {
        backend be none;
        sub vcl_recv {
                return (pipe);
        }
        sub vcl_pipe {
                return (synth(505, req.http.foo));
        }
} -start

client c1 {
        txreq -hdr "foo: foo"
        rxresp
        expect resp.status == 505
        expect resp.reason == foo
} -run
```

**Comment by Dridi** [comment]
I think I was correct, this minimal patch seems to do the trick for me:

```diff
--- i/bin/varnishd/cache/cache_req_fsm.c
+++ w/bin/varnishd/cache/cache_req_fsm.c
@@ -735,6 +735,12 @@ cnt_pipe(struct worker *wrk, struct req *req)
        switch (wrk->handling) {
        case VCL_RET_SYNTH:
                req->req_step = R_STP_SYNTH;
+               if (req->err_reason != NULL &&
+                   !WS_Inside(req->ws, req->err_reason, NULL)) {
+                       req->err_reason = WS_Copy(req->ws, req->err_reason, -1);
+                       if (req->err_reason == NULL)
+                               req->req_step = R_STP_VCLFAIL; /* XXX: log something? */
+               }
                nxt = REQ_FSM_MORE;
                break;
        case VCL_RET_PIPE:
```

Needs further research.

**Comment by nigoroll** [comment]
fixed via #3470
